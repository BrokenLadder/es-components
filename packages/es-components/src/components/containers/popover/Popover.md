```
const styles = {
    display: 'inline-block',
    margin: '10px'
};

<div>
    <div style={styles}>
    <Popover
        name="topEx"
        title="Top"
        content="This is the popover's content. 3 wolf moon officia aute, non cupidatat skateboard dolor brunch."
        placement="top"
    >
        Popover on Top
    </Popover>
    </div>

    <div style={styles}>
    <Popover
        name="bottomEx"
        title="Bottom"
        content="This is the popover's content. 3 wolf moon officia aute, non cupidatat skateboard dolor brunch."
    >
        Popover on Bottom
    </Popover>
    </div>

    <div style={styles}>
    <Popover
        name="leftEx"
        title="Left"
        content="This is the popover's content. 3 wolf moon officia aute, non cupidatat skateboard dolor brunch."
        placement="left"
    >
        Popover on Left
    </Popover>
    </div>

    <div style={styles}>
    <Popover
        name="rightEx"
        title="Right"
        content="This is the popover's content. 3 wolf moon officia aute, non cupidatat skateboard dolor brunch."
        placement="right"
    >
        Popover on Right
    </Popover>
    </div>
    
</div>
```

```
<div>
Here's an example of a
<Popover
    name="popEx"
    title="Popover"
    content="This is the popover's content. 3 wolf moon officia aute, non cupidatat skateboard dolor brunch."
    isLinkButton
    hasCloseButton
>
    Popover
</Popover>
styled like a link.
</div>
```

Popovers can wrap Icons or other components. These wrapped elements should include the `ariaLabel` prop for screen readers to identify them.
```
<div>
Click the icon for some help text.
<Popover
    name="iconEx"
    title="Icon Popover"
    content="Here's some help text!"
    suppressUnderline
    ariaLabel="Icon"
    isLinkButton
>
    <Icon name='question-circle' />
</Popover>
</div>
```

```
const styles = {
    display: 'inline-block',
    margin: '0 10px'
};

<div>
    <div style={styles}>
    <Popover
        name="smEx"
        title="Small Arrow"
        content="This is the popover's content."
        placement="top"
        isOutline
        arrowSize="sm"
    >
        Small Arrow
    </Popover>
    </div>

    <div style={styles}>
    <Popover
        name="dfltEx"
        title="Default Arrow"
        content="This is the popover's content."
        isOutline
    >
        Default Arrow
    </Popover>
    </div>

    <div style={styles}>
    <Popover
        name="lgEx"
        title="Large Arrow"
        content="This is the popover's content."
        placement="top"
        isOutline
        arrowSize="lg"
    >
        Large Arrow
    </Popover>
    </div>

    <div style={styles}>
    <Popover
        name="noneEx"
        title="No Arrow"
        content="This is the popover's content."
        placement="bottom"
        isOutline
        arrowSize="none"
    >
        No Arrow
    </Popover>
    </div>
</div>
```
