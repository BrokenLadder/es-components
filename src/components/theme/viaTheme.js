const white = '#fff';
const gray1 = '#f4f4f4';
const gray2 = '#ecedee';
const gray3 = '#e4e4e4';
const gray4 = '#d8d8d8';
const gray5 = '#c5c5c5';
const gray6 = '#979797';
// const gray7 = '#707070';
const gray8 = '#585858';
const gray9 = '#444';
const black = '#000';

const grayLightest = gray1;
const grayLighter = gray2;
const grayLight = gray3;
const gray = gray5;
const grayDark = gray6;
const grayDarker = gray8;
const grayDarkest = gray9;

const dflt = '#d8d8d8';
const dfltHover = '#bfbfbf';
const primary = '#007fa7';
const primaryHover = '#005874';
const info = '#069';
const infoHover = '#004466';
const success = '#298544';
const successHover = '#1d5e30';
const warning = '#b35f00';
const warningHover = '#804400';
const danger = '#c00';
const dangerHover = '#990000';
const advisor = warning;

const boxShadowLight = 'rgba(0, 0, 0, 0.075)';
const boxShadowDark = 'rgba(0, 0, 0, 0.5)';

const inputBorder = gray4;
const inputFocus = '#83bffc';
const inputBoxShadow = 'rgba(102, 175, 233, 0.6)';

// Datepicker colors
const navArrow = gray;
const navArrowHover = grayDark;
const selected = '#216ba5';
const hover = '#1d5d90';
const keyboard = '#2a87d0';
const inRange = 'rgba(33, 107, 165, 0.5)';
const highlight = '#3dcc4a';
const highlightHover = '#32be3f';

const theme = {
  colors: {
    black,
    grayDarkest,
    grayDarker,
    grayDark,
    gray,
    grayLight,
    grayLighter,
    grayLightest,
    white,
    dflt,
    dfltHover,
    primary,
    primaryHover,
    success,
    successHover,
    info,
    infoHover,
    warning,
    warningHover,
    danger,
    dangerHover,
    advisor,
    boxShadowLight,
    boxShadowDark,
    inputFocus
  },
  datepickerColors: {
    navArrow,
    navArrowHover,
    selected,
    hover,
    keyboard,
    inRange,
    highlight,
    highlightHover
  },
  buttonStyles: {
    buttonsNormal: {
      primary: {
        bgColor: primary,
        textColor: white,
        hoverBgColor: primaryHover,
        hoverTextColor: white,
        activeBgColor: primaryHover,
        activeTextColor: white,
        boxShadowColor: primaryHover
      },
      secondary: {
        bgColor: dflt,
        textColor: white,
        hoverBgColor: dfltHover,
        hoverTextColor: white,
        activeBgColor: dfltHover,
        activeTextColor: white,
        boxShadowColor: dfltHover
      },
      success: {
        bgColor: success,
        textColor: white,
        hoverBgColor: successHover,
        hoverTextColor: white,
        activeBgColor: successHover,
        activeTextColor: white,
        boxShadowColor: successHover
      },
      info: {
        bgColor: info,
        textColor: white,
        hoverBgColor: infoHover,
        hoverTextColor: white,
        activeBgColor: infoHover,
        activeTextColor: white,
        boxShadowColor: infoHover
      },
      warning: {
        bgColor: warning,
        textColor: white,
        hoverBgColor: warningHover,
        hoverTextColor: white,
        activeBgColor: warningHover,
        activeTextColor: white,
        boxShadowColor: warningHover
      },
      danger: {
        bgColor: danger,
        textColor: white,
        hoverBgColor: dangerHover,
        hoverTextColor: white,
        activeBgColor: dangerHover,
        activeTextColor: white,
        boxShadowColor: dangerHover
      }
    },
    buttonsAlt: {
      primary: {
        bgColor: white,
        textColor: primary,
        hoverBgColor: primary,
        hoverTextColor: white,
        activeBgColor: primaryHover,
        activeTextColor: white,
        borderColor: primary
      },
      gray: {
        bgColor: white,
        textColor: '#63666a',
        hoverBgColor: '#63666a',
        hoverTextColor: white,
        activeBgColor: '#4a4d50',
        activeTextColor: white,
        borderColor: '#63666a'
      },
      magenta: {
        bgColor: white,
        textColor: '#c110a0',
        hoverBgColor: '#c110a0',
        hoverTextColor: white,
        activeBgColor: '#920C79',
        activeTextColor: white,
        borderColor: '#c110a0'
      },
      violet: {
        bgColor: white,
        textColor: '#702082',
        hoverBgColor: '#702082',
        hoverTextColor: white,
        activeBgColor: '#4d1659',
        activeTextColor: white,
        borderColor: '#702082'
      }
    }
  },
  // buttonSizes should always have default, lg, sm, xs
  buttonSizes: {
    lg: {
      borderRadius: '3px',
      fontSize: '26px',
      padding: '8px 20px 5px'
    },
    default: {
      borderRadius: '2px',
      fontSize: '18px',
      padding: '5px 15px 4px'
    },
    sm: {
      borderRadius: '2px',
      fontSize: '14px',
      padding: '4px 10px 3px'
    },
    xs: {
      borderRadius: '2px',
      fontSize: '12px',
      padding: '3px 10px 2px'
    }
  },
  validationIconName: {
    success: 'ok',
    warning: 'warning-sign',
    danger: 'remove'
  },
  validationTextColor: {
    success,
    warning: warningHover,
    danger
  },
  validationInputColor: {
    success: {
      borderColor: success,
      boxShadow: `inset 0 0 5px ${success}`,
      focusBorderColor: success,
      focusBoxShadow: `0 0 6px ${success}`
    },
    warning: {
      borderColor: warning,
      boxShadow: `inset 0 0 5px ${warning}`,
      focusBorderColor: warning,
      focusBoxShadow: `0 0 6px ${warning}`
    },
    danger: {
      borderColor: danger,
      boxShadow: `inset 0 0 5px ${danger}`,
      focusBorderColor: danger,
      focusBoxShadow: `0 0 6px ${danger}`
    },
    default: {
      borderColor: inputBorder,
      boxShadow: `inset 0 1px 1px ${boxShadowLight}`,
      focusBorderColor: inputFocus,
      focusBoxShadow: `0 0 8px ${inputBoxShadow}`
    }
  },
  sizes: {
    baseFontSize: 18,
    baseLineHeight: 24
  },
  screenSize: {
    retina: '320px',
    phone: '480px',
    tablet: '768px',
    desktop: '992px',
    widescreen: '1200px'
  }
};

export default theme;
