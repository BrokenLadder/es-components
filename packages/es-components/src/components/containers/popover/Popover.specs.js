/* eslint-env jest */

import React from 'react';
import { render, cleanup, waitForElement } from 'react-testing-library';
import { ThemeProvider } from 'styled-components';
import viaTheme from 'es-components-via-theme';

import Popover from './Popover';

beforeEach(cleanup);

function buildPopover(props) {
  const defaults = {
    name: 'popTest',
    title: 'Popover Title',
    content: 'This is the popover content.'
  };
  const mergedProps = Object.assign({}, defaults, props);
  return (
    <ThemeProvider theme={viaTheme}>
      <Popover {...mergedProps}>Popover Text</Popover>
    </ThemeProvider>
  );
}

it('can be toggled by clicking the button', async () => {
  const { getByText, queryByText } = render(buildPopover());
  const trigger = getByText('Popover Text');

  trigger.click();
  const popoverContent = queryByText('This is the popover content.');
  expect(popoverContent).not.toBeVisible();

  trigger.click();
  expect(popoverContent).not.toBeVisible();
});

it('renders the title when provided', async () => {
  const { getByText, queryByText } = render(buildPopover());
  getByText('Popover Text').click();
  await waitForElement(() => getByText('Popover Title'));
  expect(queryByText('Popover Title')).not.toBeNull();
});

it('can be closed using the close button', async () => {
  const { getByText, queryByText } = render(
    buildPopover({ hasCloseButton: true })
  );
  getByText('Popover Text').click();

  const popoverContent = () => queryByText('This is the popover content.');
  popoverContent()
    .parentElement.querySelector('button')
    .click();

  await waitForElement(() => popoverContent() === null);
  expect(popoverContent()).toBeNull();
});

it('can be closed using the alternative close button', async () => {
  const { getByText, queryByText } = render(
    buildPopover({ hasAltCloseButton: true })
  );
  getByText('Popover Text').click();
  getByText('Popover Title')
    .parentElement.querySelector('button')
    .click();

  await waitForElement(
    () => queryByText('This is the popover content.') === null
  );
  expect(queryByText('This is the popover content.')).toBeNull();
});

it('sets focus on a focusable element within the content', () => {
  jest.useFakeTimers();
  const popoverContent = <a href="#test">Test link</a>;
  const { getByText } = render(buildPopover({ content: popoverContent }));
  getByText('Popover Text').click();
  jest.runOnlyPendingTimers();

  expect(getByText('Test link')).toHaveFocus();
});
