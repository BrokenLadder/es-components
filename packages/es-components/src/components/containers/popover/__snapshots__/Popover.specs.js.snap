// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`popover component renders as expected 1`] = `
<div
  className="sc-bZQynM dSsQJY"
>
  <div>
    <div>
      <button
        aria-expanded={false}
        className="sc-gzVnrw jgosdJ sc-htpNat cJWpVN sc-bwzfXH hXCWXw"
        onClick={[Function]}
      >
        <span
          aria-hidden={false}
        >
          Popover Text
        </span>
        <span
          className="sc-htoDjs iqIjis"
        />
      </button>
    </div>
    <div
      className="sc-EHOje vChbq"
      name="popTest"
    >
      <div
        className="popTest-popper"
        data-placement={undefined}
        data-x-out-of-boundaries={undefined}
        style={
          Object {
            "opacity": 0,
            "pointerEvents": "none",
            "position": "absolute",
            "transition": "opacity 0.2s ease-in-out",
            "zIndex": 5,
          }
        }
      >
        <div
          className="sc-dnqmqq BnNnH"
          role="dialog"
        >
          <div
            className="sc-iwsKbI cirLJO"
            tabIndex={-1}
          >
            <h3
              className="sc-gZMcBi cPBpHt"
            >
              Popover Title
            </h3>
          </div>
          <div
            className="sc-gqjmRU eiaKpk"
          >
            <div
              className="sc-VigVT cztEQg"
            >
              This is the popover content.
            </div>
          </div>
        </div>
        <span
          className="popTest-popper__arrow"
          style={Object {}}
        />
      </div>
    </div>
  </div>
</div>
`;
